plugins/redis/redisasync.cpp

```diff
@@ -0,0 +1,870 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2014 HPCC Systems.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+#include "hiredis/adapters/libev.h"
+
+#include "platform.h"
+#include "eclrtl.hpp"
+#include "jstring.hpp"
+#include "jsem.hpp"
+#include "jsocket.hpp"
+#include "jthread.hpp"
+#include "jmutex.hpp"
+#include "redisplugin.hpp"
+#include "redissync.hpp"
+#include "redisasync.hpp"
+
+#include "hiredis/async.h"
+
+namespace RedisPlugin
+{
+static CriticalSection crit;
+static const char * REDIS_LOCK_PREFIX = "redis_ecl_lock";// needs to be a large random value uniquely individual per client
+static const unsigned REDIS_LOCK_EXPIRE = 60; //(secs)
+static const unsigned REDIS_MAX_LOCKS = 9999;
+static const unsigned TIMEOUT = 2000;//ms
+static const ev_tstamp EV_TIMEOUT = TIMEOUT/1000.;//secs
+class SubContainer;
+
+class KeyLock : public CInterface
+{
+public :
+    KeyLock(ICodeContext * ctx, const char * _options, const char * _key, const char * _channel, unsigned __int64 _database);
+    ~KeyLock();
+
+    inline const char * getKey()     const { return key.str(); }
+    inline const char * getOptions() const { return options.str(); }
+    inline const char * getChannel()  const { return channel.str(); }
+    inline unsigned __int64 getDatabase()  const { return database; }
+
+private :
+    StringAttr options; //shouldn't be needed, tidy isSameConnection to pass 'master' & 'port'
+    StringAttr master;
+    int port;
+    unsigned __int64 database;
+    StringAttr key;
+    StringAttr channel;
+};
+KeyLock::KeyLock(ICodeContext * ctx, const char * _options, const char * _key, const char * _channel, unsigned __int64 _database)
+{
+    options.set(_options);
+    key.set(_key);
+    channel.set(_channel);
+    database =_database;
+    RedisPlugin::parseOptions(ctx, _options, master, port);
```

Comment:
It wasn't immediately clear that master/port were set up by this function.  Not sure how you would make it clearer.
One possibility would be to have an RedisOptions class containing master and port and call options.parse(ctx, _options).  Ok as it is though.